### ISSUE TITLE
Loss is NaN  when using Turicreate 5.0

### ISSUE THREAD

**Issue by silvywilson** [apple/turicreate#1041]
Turicreate returns 'NaN' for loss after the first iteration while using GPU. When changing to CPU it runs correctly. The OS version is 10.14 (Mojave). GPU is AMD Radeon 455 (2048 MB). Batch size is 16. I tried changing the batch size to 8, but it did not take it. The Mac is a MacBook Pro with 16 GB RAM.
Turicreate version is 5.0 (released 2 days ago - 22nd August, 2018).

**Comment by TobyRoseman** [comment]
@silvywilson - can you tell us a bit more about your use case. What type of model are you trying to train?

**Comment by silvywilson** [comment]
We are using ObjectDetection model.  We have about 50 images with three labels in each image in the training data set.


________________________________

**Comment by TobyRoseman** [comment]
@silvywilson - 50 images is a very small data set for object detection. That could be the cause of this problem.

Are you able to share your data? Or can you reproduce the behavior using our [Introductory Example](https://apple.github.io/turicreate/docs/userguide/object_detection/)?

**Comment by silvywilson** [comment]
Hi Toby,

I would agree with you except that when we train the same dataset using CPU instead of GPU, we actually get the results.  We want to be able to use GPU since the run time is considerably less.  With using CPU, we created the model and ran the test set and was able to get results.

Silvy

**Comment by cianiandreadev** [comment]
It could be due the fact that using CPU you may have a `batch_size` of at least 32 while with GPU it may not using the same value. Check this parameter and eventually try to set to 32 also for the GPU.
You can find the reference in [the doc](https://apple.github.io/turicreate/docs/api/generated/turicreate.object_detector.create.html).

But also as @TobyRoseman pointed, 50 images are too less for the training.

**Comment by silvywilson** [comment]
Hi Andrea,

I was under the impression that a higher batch size took up more memory.  So I had tried setting it to 8.  But the console print out said ‘ setting batch_size to 16’ and it did not help.  So I thought it did not take the batch size value.  I took your advice and set the batch size to 32 and even though it still printed out ‘Setting batch size to 16’,  it looks like I am not getting ‘NaN’ any more (at least not so far).  Let us see.  So the print out must be wrong!
And thanks for the help.

**Comment by silvywilson** [comment]
It finished running with no problems . and it was much faster. Thanks.


 If batch size is set to 32, but it prints out 16, wouldn't it be a bug in the turicreate code?


How much video memory would you recommend?


Silvy

________________________________

**Comment by cianiandreadev** [comment]
Well, it's hard to say. The MBP can sometimes have shared GPU memory or some of them can be occupied with some other process (like X rendering). I'm using a Linux PC with no UI at all with a GTX 970 (4 GB of VRAM) and I could train with a batch size of 32. But I noticed that if there is not enough memory when I set the batch size to 32 it is put back to 16 before training by the system (I wouldn't call it bug, since otherwise would probably crash). 

What I would suggest you as alternative is to use CPU to make initial tests (with a iteration limit of ~200) to see if the loss lowers down in these iterations (with 200 iterations should already low enough to see that is training well), then I would move everything to an (example) Amazon GPU service and train everything there **without** setting any iteration limits (Turi Create has already a very good algorithm to guess the correct amount of iterations based on the input you have).

**Comment by silvywilson** [comment]
When I set the batch size to 32 (with GPU) in my script, the loss decreases with the number of iterations.  I set it to 2000 and it worked fine (loss continue to decrease) for all iterations.  The message showed the batch size is set to 16 still.  I will probably run it w/o setting an iteration number so that the app can pick the right number.  If I do not set the batch size, the loss becomes ‘NaN’.  Since I am suspecting that the message displayed may not be reflecting the actual batch size, I cannot say what the default batch size is.

When using CPU, it automatically sets the batch size to 32 and the message shows the correct batch size.  If I change the batch size to 16, it shows batch size is set to 16.  Essentially, the batch size displayed is the same as what is set in my python script.

**Comment by cianiandreadev** [comment]
@silvywilson Maybe this weird behavior of the batch size settings can be reported as a separate bug. Just do a quick look if there isn't something reported about this already, and eventually create a separate bug? Because at this point the title of this issue is no longer really precise for this behavior.

**Comment by silvywilson** [comment]
Ok - I can do that.  Thanks for all your help.

Silvy


________________________________

**Comment by srikris** [comment]
Closing this right now. Please re-open if we have a repro.

**Comment by silvywilson** [comment]
Ok
